Added support for mutually recursive function and predicate definitions

hra687261 opened the pull request:
A proposition to fix the issue #548
I chose to add the `with` keyword as suggested by @claudemarche.
I don't see why a predicate can't be mutually recursive with a function and vice-versa, therefore I added one constructor for both.
Of course for now Dolmen doesn't support this, therefore the CI won't pass

hra687261 committed:
Added support for mutually recursive fun/pred defs

Halbaroth commented:
Great! It's quite strange that the feature has not been added before since everything works well with such a modification. 
Besides, some people use custom parsers. Modifying the legacy parser API could break temporary their stuff.  It would be better to only add this feature in Dolmen and tell Claude to use the Dolmen front-end now.

Gbury commented:
one alternative syntax option would be to require something like `function f(x) : ret = ... and function g(y) : ... = ...`, i.e. the addition of the `function` keyword at each function definition should be enough to make the parsing non ambiguous, and not break anything that is currently accepted.

hra687261 commented:
How does this addition break anything? What am I missing? The `with` keyword was just added and didn't exist before and mutually recursive functions where simply not supported. This addition doesn't change anything to the definition of non-mutually-recursive functions and predicates.

Although in a purely subjective manner and from an aesthetic point of view I find the `with` keyword more "elegant" than `and function ...`/`and predicate ...`, I don't really have a strong opinion on the subject and don't mind changing it. 

The only ambiguity I see is that there is no distinction between predicates and functions that return bools, but semantically that doesn't change anything internally AFAIK.
https://github.com/OCamlPro/alt-ergo/blob/5f63a49c793e777f11d12228ed2aa3195d5fbbb5/src/lib/frontend/typechecker.ml#L2228-L2242

Gbury commented:
> How does this addition break anything? What am I missing? The `with` keyword was just added and didn't exist before and mutually recursive functions where simply not supported. This addition doesn't change anything to the definition of non-mutually-recursive functions and predicates.

Adding a keyword is breaking in the sense that before, `with` was a valid name for a variable and/or a function, but once it is a keyword, it may not be used as a variable name anymore (changing the set of keywords of a language is almost always breaking).

hra687261 commented:
True, I'll change it.

hra687261 committed:
Use  `and function|predicate` instead of `with`
for mutually recursive function definitions

Halbaroth commented:
I prefer the Guillaume's syntax but I haven't a strong opinion too. In my opinion, users expect some changes when we will deprecate the legacy parser but we should'nt break the legacy parser before this change ;)

Halbaroth commented:
If your PR is ready for reviewing, let me know and I'll review it right now ;)

Gbury commented on src/lib/frontend/typechecker.ml:
```diff
@@ -2154,6 +2156,51 @@ let type_user_defined_type_body ~is_recursive env acc (loc, ls, s, body) =
     assert (not is_recursive); (* Abstract types are not recursive *)
     acc, env
 
+let type_fun_aux_1 env loc n ?ret_ty l  =
```
is it possible to find a better name for this function (and for `type_fun_aux_2`) ? something that would better describe what the function is doing...
Something along the lines of `declare` for `aux_1` and define for `aux_2` if I understand the code correctly.

hra687261 replied:
Done.

Gbury commented on src/parsers/native_parser.mly:
```diff
@@ -229,11 +238,21 @@ algebraic_args:
+      others = and_recursive_ty_opt
+      { (($startpos, $endpos), ty_vars, ty, enum) :: others}
+
+with_recursive_def_opt:
```
```suggestion
and_recursive_def_opt:
```

hra687261 replied:
Done.

hra687261 commented:
@Halbaroth it is!

hra687261 committed:
Renaming:
- type_fun_aux_1 -> declare_fun
- type_fun_aux_2 -> define_fun
- with_recursive_def_opt -> and_recursive_def_opt
